### PR TITLE
Staging

### PR DESCRIPTION
**Pull request by falonofthetower**

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -32,17 +37,6 @@ def show
...
-  private
-
-  def handle_invitation
-    if params[:invitation_token].present?
-      invitation = Invitation.where(token: params[:invitation_token]).first
-      @user.follow(invitation.inviter)
-      invitation.inviter.follow(@user)
-      invitation.update_column(:token, nil)
-    end
+    params.require(:user).permit(:email, :password, :full_name, :stripeToken)
```
Comment (tomtomecek) [inline app/controllers/users_controller.rb]: Why is `stripeToken` among required params. Why do you think it figures under `:user` as well?

File: app/services/user_signup.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        self
+      else
+        @status = :failure
+        @error_message = charge.error_message
+        self
+      end
+    else
+      @status = :failure
+      @error_message = "Invalid user info, please fix highlighted fields"
+      self
```
Comment (tomtomecek) [inline app/services/user_signup.rb]: All 3 branches return `self`. Why not have whole method return `self` instead :)

File: app/services/user_signup.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+  def successful?
+    @status == :success
+  end
+
+  private
+
+  def handle_invitation(invitation_token)
+    if invitation_token.present?
+      invitation = Invitation.where(token: invitation_token).first
```
Comment (tomtomecek) [inline app/services/user_signup.rb]: `where(args).first` should be `find_by`. I wonder what was the reason you have not changed this earlier.

File: spec/controllers/users_controller_spec.rb [hunk 4]
```diff
@@ -9,109 +9,38 @@
...
+          :sign_up_result,
+          successful?: false,
+          error_message: "This is failure"
+        )
+        UserSignup.any_instance.should_receive(:sign_up).and_return(result)
+        post :create, user: Fabricate.attributes_for(:user)
       end
 
-      it "does not create user" do
+      it "does not create a new user record" do
```
Comment (tomtomecek) [inline spec/controllers/users_controller_spec.rb]: This is no longer responsibility of `users_controller`. You should not test whether user record is created here, `UserSignUp` is now the owner of this process.

File: spec/controllers/users_controller_spec.rb [hunk 5]
```diff
@@ -9,109 +9,38 @@
...
 
-      it "renders the :new template" do
+      it "renders the new template" do
         expect(response).to render_template :new
       end
 
-      it "sets @user" do
-        expect(assigns(:user)).to be_instance_of(User)
+      it "sets the flash error message" do
+        expect(flash[:danger]).to eq("This is failure")
```
Comment (tomtomecek) [inline spec/controllers/users_controller_spec.rb:39]: Testing equal flash messages in controller makes your specs/tests brittle. If so these should be tested in feature specs or at the very best at models.

File: spec/features/invite_user_sign_up_spec.rb [hunk 6]
```diff
@@ -1,7 +1,7 @@
 require "spec_helper"
 
 feature "user invites another user and they sign up" do
-  scenario "user sends invite, invitee signs up" do
+  scenario "user sends invite, invitee signs up", js: true, vcr: true do
```
Comment (tomtomecek) [inline spec/features/invite_user_sign_up_spec.rb]: I prefer to write `scenario "...", :js, :vcr do`

File: spec/models/stripe_wrapper_spec.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
+require "spec_helper"
+
+describe StripeWrapper do
+  describe StripeWrapper::Charge do
+    describe ".create" do
+      it "makes a successful charge", :vcr do
```
Comment (tomtomecek) [inline spec/models/stripe_wrapper_spec.rb]: `:vcr` tag could be refactored to the level higher, as it is @ every spec anyway.

File: spec/services/user_signup_spec.rb [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+            :user,
+            email: "hans@falcon.com",
+            full_name: "Hans Solo",
+            password: "password"
+          )
+        ).sign_up("stripe_token", invitation.token)
+      end
+
+      it "makes the user follow the inviter" do
+        hans = User.where(email: "hans@falcon.com").first
```
Comment (tomtomecek) [inline spec/services/user_signup_spec.rb]: `find_by` instead of `where.first`

File: spec/controllers/users_controller_spec.rb (hunk 5 above)
Comment (falonofthetower) [inline spec/controllers/users_controller_spec.rb:39]: Actually it is only testing that the flash message is set according to what the error_message is. That logic is in fact controlled within the user controller.

``` ruby
      @error_message = charge.error_message
```

Comment (tomtomecek) [inline spec/controllers/users_controller_spec.rb:39]: testing that the flash message is set is rather tested `expect(flash[:danger]).to be_present` or other presence matcher.

**Pull request merged**